security/src/main/scala/pl/touk/nussknacker/ui/security/oauth2/JwtValidator.scala

```diff
@@ -1,26 +1,27 @@
 package pl.touk.nussknacker.ui.security.oauth2
 
-import java.security.PublicKey
-import cats.data.ValidatedNel
+import cats.data.Validated
 import cats.implicits._
 import io.circe.Decoder
 import pdi.jwt.{JwtCirce, JwtOptions}
-import pl.touk.nussknacker.ui.security.oauth2.OAuth2ErrorHandler.{OAuth2Error, OAuth2JwtError}
+import pl.touk.nussknacker.ui.security.oauth2.OAuth2ErrorHandler.{OAuth2JwtDecodeClaimsError, OAuth2JwtDecodeClaimsJsonError, OAuth2JwtDecodeRawError, OAuth2JwtKeyDetermineError, ParsedToken}
 
-import scala.util.{Failure, Success, Try}
+import java.security.PublicKey
+import scala.util.Try
 
 class JwtValidator(keyProvider: Option[String] => PublicKey) {
-  def introspect[TokenClaims: Decoder](token: String): ValidatedNel[OAuth2ErrorHandler.OAuth2Error, TokenClaims] = {
-    JwtCirce.decodeAll(token, JwtOptions.DEFAULT.copy(signature = false))
-      .flatMap { case (header, _, _) => Try(keyProvider(header.keyId)) }
-      .flatMap { publicKey => JwtCirce.decodeJson(token, publicKey) }
-    match {
-      case Success(json) => json.as[TokenClaims] match {
-        case Left(failure) => OAuth2JwtError(s"JwtValidator: failure in decoding json: ${failure.getLocalizedMessage}").invalidNel
-        case Right(profile) => profile.validNel
-      }
-      case Failure(ex) =>
-        OAuth2JwtError(s"JwtValidator: failure in JwtCirce.decodeJson: ${ex.getLocalizedMessage}").invalidNel
-    }
+
+  def introspect[TokenClaims: Decoder](token: String): Validated[OAuth2ErrorHandler.OAuth2JwtError, TokenClaims] = {
+    (for {
+      parsedToken <- parsedTokenWithoutSignature(token)
+      publicKey <- Try(keyProvider(parsedToken.header.keyId)).toEither.leftMap(OAuth2JwtKeyDetermineError(parsedToken, _))
+      tokenClaimsJson <- JwtCirce.decodeJson(token, publicKey).toEither.leftMap(OAuth2JwtDecodeClaimsError(parsedToken, publicKey, _))
+      parsedTokenClaims <- tokenClaimsJson.as[TokenClaims].leftMap(OAuth2JwtDecodeClaimsJsonError(parsedToken, publicKey, tokenClaimsJson, _))
+    } yield parsedTokenClaims).toValidated
   }
+
+  private def parsedTokenWithoutSignature(token: String) = {
+    JwtCirce.decodeAll(token, JwtOptions.DEFAULT.copy(signature = false)).toEither.leftMap(OAuth2JwtDecodeRawError(token, _)).map(ParsedToken.apply _ tupled)
```

Comment:
leftMap + map can be one fold :)

Reply:
I will change it in next PR to avoid conflicts